### PR TITLE
Fix cache exception handling

### PR DESCRIPTION
**Pull request by cdmihai**
Fixes two issues when the plugin throws an exception in Plugin.EndBuildAsync during graph builds:
- Plugin.EndBuildAsync got called twice because of a missing finally clause: https://github.com/dotnet/msbuild/compare/master...cdmihai:fixCacheExceptionHandling?expand=1#diff-2b0716a511d8f4ee690ebd5c3a59dec1e3f9a5eab4ab2a80a1018820a658accbR1946
- Plugin.EndBuildAsync got called after the graph build returned, instead of before (which allows the GraphBuildResult to be properly set as failed with the exception from Plugin.EndBuildAsync): https://github.com/dotnet/msbuild/compare/master...cdmihai:fixCacheExceptionHandling?expand=1#diff-2b0716a511d8f4ee690ebd5c3a59dec1e3f9a5eab4ab2a80a1018820a658accbR1730

### CODE REVIEW

File: src/Build/BackEnd/BuildManager/BuildManager.cs [hunk 1]
```diff
@@ -1725,68 +1725,11 @@ private void ExecuteGraphBuildScheduler(GraphBuildSubmission submission)
...
-                var graphBuildStateLock = new object();
-
-                var blockedNodes = new HashSet<ProjectGraphNode>(projectGraph.ProjectNodes);
-                var finishedNodes = new HashSet<ProjectGraphNode>(projectGraph.ProjectNodes.Count);
-                var buildingNodes = new Dictionary<BuildSubmission, ProjectGraphNode>();
-                var resultsPerNode = new Dictionary<ProjectGraphNode, BuildResult>(projectGraph.ProjectNodes.Count);
-
-                while (blockedNodes.Count > 0 || buildingNodes.Count > 0)
+                Dictionary<ProjectGraphNode, BuildResult> resultsPerNode;
+                using (cacheServiceTask.Result)
```
Comment (Forgind) [inline src/Build/BackEnd/BuildManager/BuildManager.cs:1730]: Can you explain why this changes the ordering versus when the graph build returns?

**Review by Forgind — APPROVED** [review]
Just a note that moving or refactoring a lot of code makes it much harder to review PRs, even when it's in a separate commit.

File: src/Build/BackEnd/BuildManager/BuildManager.cs (hunk 1 above)
Comment (cdmihai) [reply src/Build/BackEnd/BuildManager/BuildManager.cs:1730]: using variables get disposed at the end of the block they're in. In this case the end of the block is **after** the last line in the block whish is the call to ReportResultsToSubmission, but the cache needs to be disposed (in order to surface any issues in EndBuildAsync) **before** the call to ReportResultsToSubmission.

**Comment by cdmihai** [comment]
This PR heavily conflicts with https://github.com/dotnet/msbuild/pull/6016, so I rebased it on #6016. Merge that one into master first.

File: src/Build/BackEnd/BuildManager/BuildManager.cs [hunk 2]
```diff
@@ -1945,8 +1945,14 @@ public void Dispose()
 
                 lock (_buildManager._syncLock)
                 {
-                    _buildManager._projectCacheService?.Result.ShutDown().GetAwaiter().GetResult();
-                    _buildManager._projectCacheService = null;
+                    try
+                    {
+                        _buildManager._projectCacheService?.Result.ShutDown().GetAwaiter().GetResult();
```
Comment (rokonec) [inline src/Build/BackEnd/BuildManager/BuildManager.cs:1950]: Nit: Is it guaranteed that it will not get in race condition with line `var projectCacheShutdown = _projectCacheService?.Result.ShutDown();` ?
I believe having protection against double execution inside `Shutdown()` method would be beneficial.

**Review by rokonec — COMMENTED** [review]
Looks OK.

File: src/Build/BackEnd/BuildManager/BuildManager.cs (hunk 2 above)
Comment (cdmihai) [reply src/Build/BackEnd/BuildManager/BuildManager.cs:1950]: I don't think it will race because EndBuild [waits for all submissions to finish](https://github.com/dotnet/msbuild/blob/6819f7ab06c3f43e83ff4059d417395e0af06c01/src/Build/BackEnd/BuildManager/BuildManager.cs#L776-L783) before disposing the plugin. This means it waits for any graph build submissions too, and those will shutdown the cache and set it to null before exiting.

---
In reply to: [583598558](https://github.com/dotnet/msbuild/pull/6110#discussion_r583598558) [](ancestors = 583598558)
